File: crates/oxc_linter/src/rules/typescript/prefer_regexp_exec.rs
use oxc_macros::declare_oxc_lint;

use crate::rule::Rule;

#[derive(Debug, Default, Clone)]
pub struct PreferRegexpExec;

declare_oxc_lint!(
    /// ### What it does
    ///
    /// Prefer `RegExp#exec()` over `String#match()` when extracting a regex match.
    ///
    /// ### Why is this bad?
    ///
    /// `exec()` is more explicit about matching with a regular expression and avoids the
    /// overloaded behavior of `String#match()`.
    ///
    /// ### Examples
    ///
    /// Examples of **incorrect** code for this rule:
    /// ```ts
    /// const text = 'value';
    /// text.match(/v/);
    /// ```
    ///
    /// Examples of **correct** code for this rule:
    /// ```ts
    /// const text = 'value';
    /// /v/.exec(text);
    /// ```
    PreferRegexpExec(tsgolint),
    typescript,
    nursery,
);

impl Rule for PreferRegexpExec {}

File: apps/oxlint/fixtures/cli/tsgolint/prefer-regexp-exec.ts
// Examples of incorrect code for prefer-regexp-exec rule

const text = 'something';
text.match(/thing/); // should prefer RegExp#exec

const search = 'thing';
text.match(search); // should prefer RegExp#exec

function countMatches(str: string, re: RegExp): number {
  return str.match(re)?.length ?? 0; // valid: unknown global flag
}
